### PR TITLE
Extracting api server code into a library: part 1

### PR DESCRIPTION
**Pull request by nikhiljindal**
Ref https://github.com/kubernetes/kubernetes/issues/2742

Created a genericapiserver package to move all generic api server code there.

Have moved the config and some setup code from masters.go there.

In future PRs will move other code (like apiserver.go, api_installer.go, etc)

cc @lavalamp

### CODE REVIEW

**Comment by lavalamp** [comment]
> kubeapiserver package

This should be the name for the kube-specific stuff, no?

(I haven't looked yet-- but pkg/master is where kube-apiserver specific stuff was originally supposed to go.)

**Comment by nikhiljindal** [comment]
As discussed will rename the package name to "genericapiserver"

**Comment by nikhiljindal** [comment]
Rebased and renamed.

File: cmd/kube-apiserver/app/server.go [hunk 1]
```diff
@@ -551,37 +552,39 @@ func (s *APIServer) Run(_ []string) error {
...
-		MinRequestTimeout:         s.MinRequestTimeout,
-		ProxyDialer:               proxyDialerFn,
-		ProxyTLSClientConfig:      proxyTLSClientConfig,
-		Tunneler:                  tunneler,
-		ServiceNodePortRange:      s.ServiceNodePortRange,
-		KubernetesServiceNodePort: s.KubernetesServiceNodePort,
+		Config: &genericapiserver.Config{
+			StorageDestinations:       storageDestinations,
+			StorageVersions:           storageVersions,
+			EventTTL:                  s.EventTTL,
```
Comment (lavalamp) [inline cmd/kube-apiserver/app/server.go]: Is this generic?

File: cmd/kube-apiserver/app/server.go [hunk 2]
```diff
@@ -551,37 +552,39 @@ func (s *APIServer) Run(_ []string) error {
...
+			KubeletClient:             kubeletClient,
```
Comment (lavalamp) [inline cmd/kube-apiserver/app/server.go]: Is this generic?

File: cmd/kube-apiserver/app/server.go [hunk 3]
```diff
@@ -551,37 +552,39 @@ func (s *APIServer) Run(_ []string) error {
...
+			ServiceClusterIPRange:     &n,
```
Comment (lavalamp) [inline cmd/kube-apiserver/app/server.go]: Is this generic?

File: cmd/kube-apiserver/app/server.go [hunk 4]
```diff
@@ -551,37 +552,39 @@ func (s *APIServer) Run(_ []string) error {
...
+			EnableLogsSupport:         s.EnableLogsSupport,
+			EnableUISupport:           true,
+			EnableSwaggerSupport:      true,
+			EnableProfiling:           s.EnableProfiling,
+			EnableWatchCache:          s.EnableWatchCache,
+			EnableIndex:               true,
+			APIPrefix:                 s.APIPrefix,
+			APIGroupPrefix:            s.APIGroupPrefix,
```
Comment (lavalamp) [inline cmd/kube-apiserver/app/server.go]: Do we really want to let people customize these two fields? APIPrefix is definitely not generic.

File: cmd/kube-apiserver/app/server.go [hunk 5]
```diff
@@ -551,37 +552,39 @@ func (s *APIServer) Run(_ []string) error {
...
+			AdmissionControl:          admissionController,
+			APIGroupVersionOverrides:  apiGroupVersionOverrides,
+			MasterServiceNamespace:    s.MasterServiceNamespace,
+			ClusterName:               s.ClusterName,
+			ExternalHost:              s.ExternalHost,
+			MinRequestTimeout:         s.MinRequestTimeout,
+			ProxyDialer:               proxyDialerFn,
+			ProxyTLSClientConfig:      proxyTLSClientConfig,
+			ServiceNodePortRange:      s.ServiceNodePortRange,
+			KubernetesServiceNodePort: s.KubernetesServiceNodePort,
```
Comment (lavalamp) [inline cmd/kube-apiserver/app/server.go]: Maybe it's better to list the things I think *are* generic:
* StorageDestinations
* StorageVersions
* Swagger
* Profiling (maybe?)
* WatchCache
* Cors
* Read/write port
* public address
* auth things
* admission - maybe?
* group version overrides
* Master service namespace -- each set of apiservers will need to make their own service
* cluster name -- do all apiservers need to know this?
* MinRequestTimeout -- maybe
* proxy stuff -- probably

Things not mentioned are not generic IMO

File: pkg/genericapiserver/genericapiserver.go [hunk 6]
```diff
@@ -0,0 +1,565 @@
...
+	ExtraServicePorts []api.ServicePort
+	// Additional ports to be exposed on the Genericapiserver endpoints
+	// Port names should align with ports defined in ExtraServicePorts
+	ExtraEndpointPorts []api.EndpointPort
+
+	KubernetesServiceNodePort int
+}
+
+// Genericapiserver contains state for a Kubernetes cluster Genericapiserver/api server.
+type Genericapiserver struct {
```
Comment (lavalamp) [inline pkg/genericapiserver/genericapiserver.go]: The type can probably be named APIServer?

File: pkg/genericapiserver/genericapiserver.go [hunk 7]
```diff
@@ -0,0 +1,565 @@
...
+
+	// Control the interval that pod, node IP, and node heath status caches
+	// expire.
+	CacheTimeout time.Duration
+
+	// The name of the cluster.
+	ClusterName string
+
+	// The range of IPs to be assigned to services with type=ClusterIP or greater
+	ServiceClusterIPRange *net.IPNet
```
Comment (lavalamp) [inline pkg/genericapiserver/genericapiserver.go]: Commented earlier on things in here that aren't generic :)

**Comment by lavalamp** [comment]
I didn't make an exhaustive pass yet.

I think you'll also need a mechanism for merging flags, so you can move generic stuff out of cmd/kube-apiserver/ -- ideally it'd just be a few lines of code to start up a totally empty apiserver.

File: cmd/kube-apiserver/app/server.go (hunk 4 above)
Comment (nikhiljindal) [inline cmd/kube-apiserver/app/server.go]: If its customizable in core master, then it needs to be customizable for all api servers.
If its not required, I can remove it everywhere.

File: cmd/kube-apiserver/app/server.go (hunk 1 above)
Comment (nikhiljindal) [inline cmd/kube-apiserver/app/server.go]: Moved from generic api server to master

File: cmd/kube-apiserver/app/server.go (hunk 2 above)
Comment (nikhiljindal) [inline cmd/kube-apiserver/app/server.go]: Moved from generic api server to master

File: cmd/kube-apiserver/app/server.go (hunk 5 above)
Comment (nikhiljindal) [inline cmd/kube-apiserver/app/server.go]: As discussed irl, refactoring cmd/kube-apiserver/app/server.go is another big change.
This PR just moves common api machinery code out from master to genericapiserver. Will handle server.go in another PR.

**Comment by nikhiljindal** [comment]
Rebased, fixed unit tests and updated code as per comments. PTAL

File: cmd/kube-apiserver/app/server.go [hunk 8]
```diff
@@ -110,7 +111,6 @@ type APIServer struct {
 	MasterServiceNamespace     string
 	RuntimeConfig              util.ConfigurationMap
 	KubeletConfig              kubeletclient.KubeletClientConfig
-	ClusterName                string
```
Comment (krousey) [inline cmd/kube-apiserver/app/server.go]: Was this just not needed? (many more dumb questions to follow)

Comment (nikhiljindal) [inline cmd/kube-apiserver/app/server.go]: Yes. Have pulled out the change in https://github.com/kubernetes/kubernetes/pull/18388.
Will rebase once that is merged.

File: pkg/genericapiserver/doc.go [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+limitations under the License.
+*/
+
+// Package genericapiserver contains code to setup a generic kubernetes-like API server.
+// This does not contain any kubernetes API specific code.
+// Note that this is a work in progress. We are pulling out generic code (specifically from
+// pkg/master and pkg/apiserver) here.
+// We plan to move this package into a separate repo on github once it is done.
+// For more details: https://github.com/kubernetes/kubernetes/issues/2742
+package genericapiserver
```
Comment (krousey) [inline pkg/genericapiserver/doc.go]: Minor nit, no need to change: Generally I would only put package documentation in a doc.go file if it was a substantial documentation or there were several files in the package that would make finding it hard. Here it's fairly short, and there's only one non-test go file.

File: pkg/genericapiserver/genericapiserver.go [hunk 10]
```diff
@@ -0,0 +1,541 @@
...
+	DefaultEtcdPathPrefix = "/registry"
+)
+
+// StorageDestinations is a mapping from API group & resource to
+// the underlying storage interfaces.
+type StorageDestinations struct {
+	APIGroups map[string]*StorageDestinationsForAPIGroup
+}
+
+type StorageDestinationsForAPIGroup struct {
```
Comment (krousey) [inline pkg/genericapiserver/genericapiserver.go]: I know you're just moving code, but does this need to be public?

File: pkg/genericapiserver/genericapiserver.go (hunk 6 above)
Comment (krousey) [inline pkg/genericapiserver/genericapiserver.go]: Agreed. At the very least API and Server should be capitalized.

File: pkg/genericapiserver/genericapiserver.go [hunk 11]
```diff
@@ -0,0 +1,541 @@
...
+	Handler         http.Handler
+	InsecureHandler http.Handler
+
+	// Used for custom proxy dialing, and proxy TLS options
+	ProxyTransport http.RoundTripper
+
+	KubernetesServiceNodePort int
+}
+
+func (s *Genericapiserver) StorageDecorator() generic.StorageDecorator {
```
Comment (krousey) [inline pkg/genericapiserver/genericapiserver.go]: Wasn't this supposed to be a method of the config?

File: pkg/master/master.go [hunk 12]
```diff
@@ -75,248 +69,36 @@ import (
...
-	publicReadWritePort  int
-	serviceReadWriteIP   net.IP
-	serviceReadWritePort int
-	masterServices       *util.Runner
-	extraServicePorts    []api.ServicePort
-	extraEndpointPorts   []api.EndpointPort
-
-	// storage contains the RESTful endpoints exposed by this master
-	storage map[string]rest.Storage
+	*genericapiserver.Genericapiserver
```
Comment (krousey) [inline pkg/master/master.go]: Why embed this? In the code below, it makes it difficult to determine if a member or method is defined on this struct or in the generic API package. Are consumers expected to have access to these members and methods too?

File: pkg/master/master_test.go [hunk 13]
```diff
@@ -90,37 +92,32 @@ func TestNew(t *testing.T) {
...
+}
+
+// TestNew verifies that the New function returns a Master
+// using the configuration properly.
+func TestNew(t *testing.T) {
+	master, etcdserver, config, assert := newMaster(t)
+	defer etcdserver.Terminate(t)
 
 	// Verify many of the variables match their config counterparts
-	assert.Equal(master.enableCoreControllers, config.EnableCoreControllers)
```
Comment (krousey) [inline pkg/master/master_test.go]: Where did this assertion go?

**Comment by krousey** [comment]
Sorry that took me a while to get through. It was hard to follow all the changes with the move.

File: pkg/genericapiserver/doc.go (hunk 9 above)
Comment (nikhiljindal) [inline pkg/genericapiserver/doc.go]: Ack. There is just one file for now. Will add more code in future PRs.
This comment is for anyone else who might be looking at the code and wondering if their code should go in this package.

File: pkg/genericapiserver/genericapiserver.go (hunk 10 above)
Comment (nikhiljindal) [inline pkg/genericapiserver/genericapiserver.go]: Yes since StorageDestinations is public (it is part of GenericApiServer.Config)

File: pkg/genericapiserver/genericapiserver.go (hunk 11 above)
Comment (nikhiljindal) [inline pkg/genericapiserver/genericapiserver.go]: I moved it to server since I wanted to call it from a method in master that did not have access to the config.
Eventually, I ended up passing config to that function for something else :) So I can move this back to config, if you think its better?

File: pkg/master/master.go (hunk 12 above)
Comment (nikhiljindal) [inline pkg/master/master.go]: 1. Embedding it made it easier to refactor.
2. Most of the public members and methods of generic api server should go away if we are able to move bulk of master.InstallAPI logic to generic apiserver (which I plan to do next).

That said, I think it makes sense to not embed genericapiserver here. Will see how big of a change that is. Is it fine to do that in a separate PR if its too big?

File: pkg/master/master_test.go (hunk 13 above)
Comment (nikhiljindal) [inline pkg/master/master_test.go]: Good catch. Added

File: pkg/genericapiserver/genericapiserver.go (hunk 6 above)
Comment (nikhiljindal) [inline pkg/genericapiserver/genericapiserver.go]: Renamed to GenericAPIServer

**Comment by nikhiljindal** [comment]
Replied to all comments and updated code appropriately.
Ready for another round.

File: pkg/master/master.go (hunk 12 above)
Comment (krousey) [inline pkg/master/master.go]: Definitely a different PR. If there's a lot more work to be done here, we can just save cleanup for the final step.
